utils/src/main/scala/za/co/absa/enceladus/utils/broadcast/LocalMappingTable.scala

```diff
@@ -83,7 +86,8 @@ object LocalMappingTable {
       (keys, if (values.length == 1) values.head else new GenericRowWithSchema(values, rowSchema))
     }).toMap
 
-    LocalMappingTable(mappingTable, outputColumns, keyTypes, valueTypes)
+    val valueType: DataType = if (outputColumns.size == 1) valueTypes.head else rowSchema
+    LocalMappingTable(mappingTable, outputColumns, keyTypes, valueType)
```

Comment:
I would restructure this piece of code - line 76-89. 
I don't like that the forking between one-value and multiple-value mapping table is done on more places. Also the code is not that obvious IMO. And the _select_ on the dataFrame might be worth to avoid. 
Suggesting this code instead (put the subfunction and the top of the method):
```
    def oneValue(seq: Seq[Any]): Any = seq.head
    def moreValues(seq: Seq[Any], schema: StructType): Any = new GenericRowWithSchema(seq.toArray, schema)
    val (valueType, valueFnc) = if (outputColumns.size == 1) {
      (valueTypes.head, oneValue _)
    } else {
      (rowSchema, moreValues(_, rowSchema))
    }
    val mappingTable = mappingTableDf.collect().map(row => {
      val keys = keyFields.map(name => row.getAs[Any](name))
      val values = targetAttributes.map(name => row.getAs[Any](name))
      (keys, valueFnc(values))
    }).toMap
```

Reply:
Using `getAs[Any](name)` looks good, but does not really work for accessing nested fields, unless a kind of recursive function is used, which I don't believe is suitable here.

Reply:
Darn, you are right. 😞 
Well it could be done with:
```
    def getRowValue(row: Row, fieldNameSegments: Seq[String]): Any = {
      fieldNameSegments match {
        case Nil => null
        case head :: Nil => row.getAs[Any](head)
        case head :: tail => getRowValue(row.getAs[Row](head), tail)
      }
    }
```
And have a function in `def SchemaUtils.fieldNameToSegments(fieldName: String): Seq[String]`

**OR**

Keep it as it is, despite not nice. `select` is not an _action_ but _transformation_ I checked, so from that point of view it's OK. Then just one point: 
`val targetColumns: Seq[Column]` is declared (line 76), but `keyFields.map(c => col(c))` is used directly (line 78). Let's do it one way or another, the mix is confusing.